theano/tensor/nnet/abstract_conv.py

```diff
@@ -177,6 +179,383 @@ def conv2d(input,
+
+def bilinear_kernel_2D(ratio, normalize=True):
+    """ This function gets a ratio
+    and returns the normalized 2D filters that can
+    be applied to any image to upsample it by the ratio
+    using bilinear interpolation.
+
+    :type ratio: integer
+    : param ratio: the ratio by which an image is upsampled
+        in the 2D space.
+
+    :type normalize: bool
+    : param normalize: indicates whether to normalize the
+    kernel or not. Default is True.
+
+    returns: the 2D kernels that can be applied to
+             any given image to upsample it by indicated ratio
+             using bilinear interpolation.
+    """
+    assert ratio > 1
+    kern_size = ratio * 2 - 1
+    kern = np.zeros((kern_size, kern_size))
+    for ii in range(-ratio + 1, ratio):
+        for jj in range(-ratio + 1, ratio):
+            kern[ii + ratio - 1, jj + ratio - 1] = (ratio - abs(ii)) * (ratio - abs(jj))
+    if normalize:
+        kern /= ratio**2
+    return kern
+
+
+def bilinear_kernel_1D(ratio, normalize=True):
+    """ This function gets a ratio
+    and returns the normalized 1D filters that can
+    be applied to any image to upsample it by the ratio
+    using bilinear interpolation.
+
+    :type ratio: integer
+    : param ratio: the ratio by which an image is upsampled
+        in the 1D space.
+
+    :type normalize: bool
+    : param normalize: indicates whether to normalize the
+    kernel or not. Default is True.
+
+    returns: the 1D kernels that can be applied to
+             any given image to upsample it by indicated ratio
+             using bilinear interpolation in one dimension.
+    """
+    assert ratio > 1
+    kern_size = ratio * 2 - 1
+    kern = np.zeros((kern_size))
+    for ii in range(-ratio + 1, ratio):
+        kern[ii + ratio - 1] = ratio - abs(ii)
+    if normalize:
+        kern /= ratio
+    return kern
+
+
+def bilinear_upsampling(input,
+                        ratio,
+                        input_shape=None,
+                        use_1D_kernel=True,
+                        repeat_first_row=True,
+                        repeat_first_col=True):
+    """This function takes a stack of 2D inputs and upsamples it (in 2D space)
+    by the ratio. The subsample is assumed to be one and
+
+    :type input: symbolic 4D tensor
+    :param input: mini-batch of feature map stacks, of shape
+        (batch size, input channels, input rows, input columns).
+        See the optional parameter ``input_shape`` that will be upsampled.
+
+    :type ratio: integer
+    : param ratio: the ratio by which the input is upsampled
+        in the 2D space.
+
+    :type input_shape: None, tuple/list of len 4 of int or
+        Constant variable.
+    :param input_shape: The shape of the input variable.
+        Optional, possibly used to choose an optimal implementation.
+        You can give ``None`` for any element of the list to specify that this
+        element is not known at compile time.
+
+    :type use_1D_kernel: bool
```

Comment:
I don't think we want to let the user choose which strategy to apply. The optimization should go on under the hood without the user being aware of it. When the filter is big doing the two upsampling separately should be faster. 

Please benchmark the two implementations with a combination of values for image size, kernel size and upsampling ratio to see if it's possible to find an heuristic to determine when to apply one or the other strategy.